### PR TITLE
feat(controller): Support any() and all() for TaskGroup in depends logic. Closes #3405

### LINKED ISSUE #3405: Allow TaskGroup nodes status to better reflect skipped nodes when using expansion
See: https://github.com/argoproj/argo/issues/3378#issuecomment-654991693

### PR DESCRIPTION
**Pull request by markterm**
If a DAG task depends on a task that uses withItems, then can depend on any or all of the children being in a status. 

Checklist:

* [X] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [X] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [X] I've signed the CLA.
* [X] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [X] My builds are green. Try syncing with master if they are not. 
* [X] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

### CODE REVIEW

**Comment by mark9white** [comment]
FYI @simster7

**Comment by simster7** [comment]
Thanks, will take a look at this over the coming days

**Comment by mark9white** [comment]
Just a little nudge @simster7 ;)

**Comment by simster7** [comment]
Hi @mark9white, thanks for the nudge and sorry for the delay. I'll take a look today

**Comment by simster7** [comment]
Hey @mark9white. Thanks for submitting this change and sorry for the delay. We agreed to support this feature, but we want to consider a different syntax.

Having `any(task.Succeeded)`, doesn't make much sense as: (1) there is nothing that distinguishes whether `task` actually expanded or not, (2) overloads the meaning of `task.Succeeded`: if it's inside `any(...)` then it refers to the expanded tasks, while if it's not it refers to the parent task, and (3) frankly it doesn't make much sense.

We want to experiment a bit with possible syntaxes. Ideas that came around include:

* Having an "expander" such as `task.ExpandedTasks` resolve to a list of individual tasks statuses. We can then leverage `expr` to evaluate on them (such as `"Succeeded" in task.ExpandedTasks`)
* Having dedicated boolean attributes (such as `task.AnySucceeded`, `task.AllFailed`)
* Others?

Thoughts?

**Comment by mark9white** [comment]
Thanks for taking a look. I think dedicated boolean attributes as you suggest, eg task.AnySucceeded, would be easiest for the users to understand and use. I could change the PR implementation to that?

**Comment by simster7** [comment]
Sure sounds good. I'm thinking that we only need to have `task.AnySucceeded` since `task.AllFailed` can be inferred from it:

```
task.AllFailed = task.Failed && !task.AnySucceeded
```

What do you think? Are there any more than we might need?

**Comment by mark9white** [comment]
I think if we want to minimise the surface area then `task.AnySucceeded` would be enough for most requirements, and I'd be fine with just implementing that.

However if we want to provide full flexibility then I would add all the permutations. For example, it's not exactly possible to calculated all failed as you suggest - eg the case where one was skipped, one failed and none succeeded. And having a direct `AllFailed` makes it easier to read Argo workflows than combinations of other conditions.

**Comment by simster7** [comment]
Sure let's start with `AnySucceeded` and `AllFailed`. Please add validation such that if a user tries to use these variables in a task which does not expand, they will get an error

**Comment by alexec** [comment]
@mark9white let me know if you need any help with this.

**Comment by mark9white** [comment]
@alexec Thanks, I should be able to get to it in a couple of weeks ..

**Review by alexec — CHANGES_REQUESTED** [review]
Changes requested

**Comment by mark9white** [comment]
@simster7 @alexec FYI I have made the requested changes.

File: Makefile [hunk 1]
```diff
@@ -47,7 +47,7 @@ CONTROLLER_IMAGE_FILE  := dist/controller-image.marker
 
 # perform static compilation
 STATIC_BUILD          ?= true
-STATIC_FILES          ?= true
+STATIC_FILES          ?= false
```
Comment (alexec) [inline Makefile:50]: revert this

Comment (markterm) [reply Makefile:50]: Done

File: workflow/controller/dag.go [hunk 2]
```diff
@@ -44,7 +44,7 @@ type dagContext struct {
 
 	// dependencies is a list of all the tasks a specific task depends on. Because dependencies are computed using regex
 	// and regex is expensive, we cache the results so that they are only computed once per operation
-	dependencies map[string][]string
+	dependencies map[string]map[string]common.DependencyType
```
Comment (simster7) [inline workflow/controller/dag.go:47]: I'm almost surely wrong, but I didn't see a place where we actually use the value of the inner map (i.e. `common.DependencyType`). Can you explain why it's necessary, point me where it's used?

**Review by simster7 — CHANGES_REQUESTED** [review]
Gave this a very quick first pass

File: workflow/common/ancestry.go [hunk 3]
```diff
@@ -38,31 +40,37 @@ type expansionMatch struct {
 	end      int
 }
 
-func GetTaskDependencies(task *wfv1.DAGTask, ctx DagContext) ([]string, string) {
+type DependencyType int
+
+const (
+	TaskDependency  DependencyType = 0
```
[the remark this answers is not in the document]
Comment (markterm) [reply workflow/common/ancestry.go:46]: 👍

File: workflow/controller/dag.go (hunk 2 above)
Comment (markterm) [reply workflow/controller/dag.go]: The issue is that dagValidationContext needs the inner map, and was getting passed in as a dagContext, so I modified dagContext to keep them matching. This wasn't ideal, I've fixed it up.

**Comment by markterm** [comment]
I don't think the build failure is related to my change.

**Comment by markterm** [comment]
@alexec  @simster7  Am checking next steps here?

**Comment by alexec** [comment]
@simster7 do you have anymore feedback on this please?

**Review by simster7 — APPROVED** [review]
Sorry for the delay. This LGTM, I just want @jessesuen to approve of the syntax. I will bring this up with him tomorrow

**Comment by simster7** [comment]
@markterm Can you merger/rebase this with master please?

**Comment by markterm** [comment]
@simster7 done

**Comment by simster7** [comment]
@markterm Could you check this box?

image


This way we can merge this branch with master on our own to merge it faster

**Comment by markterm** [comment]
I don't have that option, so I've given access to the repo which will hopefully have the same effect. I've also triggered a branch update with master.

**Pull request merged**